### PR TITLE
feat: using partition key as data selector.

### PR DESCRIPTION
**Pull request by TalkWIthKeyboard**
The traces table and logs table are so big datasets, we should make some partitions to optimize the query speed. 

Condition class is an abstraction to describe the data selector, user can define their own selector with their own partition key, we provide Datawaves selector for our optimized traces table and logs table.

Some interesting things:  we use `hash()` function in spark.sql to get the hash number of the string, but the function is implemented based on int32, it has quite different from the implementation in Python, so we need to implement it by myself.

TODO:  compatible with BigEndian system.

### CODE REVIEW

File: spark3/ethereum/condition.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+        return Conditions(
+            ctype='function',
+            address_condition=return_self_condition('to_address', 'string'),
+            address_hash_condition=return_hash_condition('address_hash', 'int'),
+            selector_condition=default_trace_function_selector_condition,
+            selector_hash_condition=default_trace_function_hash_condition
+        )
+
+    @staticmethod
+    def new_datawave_log_condition():
```
Comment (onesuper) [inline spark3/ethereum/condition.py]: I guess those Conditions better be moved outside this client library.

And let the library users (like datawaves & @raycursive ) implement them : )

File: spark3/ethereum/contract.py [hunk 2]
```diff
@@ -60,48 +57,43 @@ def event_schema(self) -> Dict[str, StructType]:
...
             raise FunctionOrEventNotInContractABI()
 
-        df = self.spark3.trace_df
-
-        if len([col_name for col_name, col_type in df.dtypes if
-                col_name == "to_address" and col_type == "string"]) != 1:
-            raise ColumnNotFoundInDataFrame("to_address(string)", df)
+        function_abi = self._get_function_abi_item_json(name)
+        df = self.spark3.trace_condition \
+            .act(self.spark3.trace_df, self.address, json.loads(function_abi))
```
Comment (raycursive) [inline spark3/ethereum/contract.py]:
```suggestion
            .act(self.spark3.trace_df, self.address, function_abi)
```

Comment (raycursive) [inline spark3/ethereum/contract.py]:
```suggestion
            .act(self.spark3.trace_df, self.address, function_abi)
```

File: spark3/ethereum/contract.py [hunk 3]
```diff
@@ -60,48 +57,43 @@ def event_schema(self) -> Dict[str, StructType]:
...
+            raise ValueError('Could not call get_event_by_name without log dataframe')
+
         schema = self.event_schema.get(name)
         if schema is None:
             raise FunctionOrEventNotInContractABI()
 
-        df = self.spark3.log_df
+        event_abi = self._get_event_abi_item_json(name)
+        df = self.spark3.log_condition \
+            .act(self.spark3.log_df, self.address, json.loads(event_abi))
```
Comment (raycursive) [inline spark3/ethereum/contract.py]:
```suggestion
            .act(self.spark3.log_df, self.address, event_abi)
```

File: spark3/ethereum/contract.py [hunk 4]
```diff
@@ -60,48 +57,43 @@ def event_schema(self) -> Dict[str, StructType]:
...
+        return self._get_abi_item_json(name, 'function')
 
-    def _get_abi_item_json(self, name) -> str:
-        abi_list = [x for x in self.abi
-                    if x['type'] in ('function', 'event') and x['name'] == name]
+    def _get_abi_item_json(self, name: str, _type: str) -> str:
+        abi_list = [x for x in self.abi if x['type'] == _type and x['name'] == name]
         if len(abi_list) == 0:
             raise FunctionOrEventNotInContractABI()
         return json.dumps(abi_list[0])
```
Comment (raycursive) [inline spark3/ethereum/contract.py]:
```suggestion
        return abi_list[0]
```

File: test/spark3/ethereum/test_contract.py [hunk 5]
```diff
@@ -86,8 +106,9 @@ def test_get_function_by_name(self):
         self.assertEqual(input['tuple']['key'], 'test')
         self.assertEqual(input['tuple']['value'], 1234)
 
+        input_schema_filed = [field for field in new_df.schema.fields if field.name == 'function_parameter'][0]
```
Comment (raycursive) [inline test/spark3/ethereum/test_contract.py]:
```suggestion
        input_schema_field = [field for field in new_df.schema.fields if field.name == 'function_parameter'][0]
```

File: test/spark3/ethereum/test_contract.py [hunk 6]
```diff
@@ -86,8 +106,9 @@ def test_get_function_by_name(self):
...
         input_schema = {field['name']: field['type'] for field in
-                        new_df.schema.fields[1].dataType.jsonValue()['fields'][0]['type']['fields']}
+                        input_schema_filed.dataType.jsonValue()['fields'][0]['type']['fields']}
```
Comment (raycursive) [inline test/spark3/ethereum/test_contract.py]:
```suggestion
                        input_schema_field.dataType.jsonValue()['fields'][0]['type']['fields']}
```

File: spark3/main.py [hunk 7]
```diff
@@ -14,15 +15,28 @@ class Spark3:
...
+                 log_condition: Optional[Conditions] = None):
         """
         :param spark: :class:`SparkSession`
         :param trace: :class:`DataFrame` to store original data to decode contract functions
         :param log: :class:`DataFrame` to store original data to decode contract events
         """
+        if trace is None and log is None:
+            raise ValueError('Either trace or log should be provided')
+
+        if trace is not None and trace_condition is None:
```
Comment (onesuper) [inline spark3/main.py]: This requirements maybe is too strict and make this client hard to use

File: spark3/ethereum/condition.py [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+            result_df = self.address_hash_condition.acts(result_df, address)
+        if self.selector_condition is not None:
+            result_df = self.selector_condition.acts(result_df, single_abi)
+        if self.selector_hash_condition is not None:
+            result_df = self.selector_hash_condition.acts(result_df, single_abi)
+
+        return result_df
+
+    @staticmethod
+    def new_datawave_trace_condition():
```
Comment (raycursive) [inline spark3/ethereum/condition.py]: maybe a better naming? (not sure if there should be datawave or datawaves / trace or traces

btw, these two staticmethods should be easily accessed by users (maybe like the way EtherscanABIProvider made?)

File: spark3/ethereum/condition.py [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+    def __init__(self,
+                 ctype: str,
+                 address_condition: Condition[str],
+                 address_hash_condition: Optional[Condition[str]] = None,
+                 selector_condition: Optional[Condition[Dict[str, any]]] = None,
+                 selector_hash_condition: Optional[Condition[Dict[str, any]]] = None):
+        if ctype != 'function' and ctype != 'event':
+            raise TypeError('The column type must be "function" or "event"')
+
+        self.ctype = ctype
```
Comment (onesuper) [inline spark3/ethereum/condition.py]: What's the purpose of this class field?

File: spark3/ethereum/condition.py [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+            key_type='int',
+            # :paras x: function abi dict
+            transform=lambda x: abs(hash_unsafe_bytes(encode_hex(function_abi_to_4byte_selector(x)))) % 10
+        )
+
+        return Conditions(
+            ctype='function',
+            address_condition=return_self_condition('to_address', 'string'),
+            address_hash_condition=return_hash_condition('address_hash', 'int'),
+            selector_condition=default_trace_function_selector_condition,
```
Comment (onesuper) [inline spark3/ethereum/condition.py]: I was wondering why we have two condition for selector for each?

File: spark3/ethereum/condition.py [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+from pyspark.sql.functions import col
+
+from spark3.exceptions import ColumnNotFoundInDataFrame
+from spark3.utils import hash_unsafe_bytes
+
+T = TypeVar('T')
+
+
+class Condition(Generic[T]):
+    def __init__(self, key_alias: str, key_type: str, transform: Callable[[T], any]):
```
Comment (onesuper) [inline spark3/ethereum/condition.py]: A straight thought: Can we also specify the return type of transform? for example, Condition[ str, int]

File: spark3/main.py (hunk 7 above)
Comment (TalkWIthKeyboard) [reply spark3/main.py:29]: move these asserts to the specific method

Comment (TalkWIthKeyboard) [reply spark3/main.py]: > This requirements maybe is too strict and make this client hard to use

move these asserts to the specific method

Comment (onesuper) [reply spark3/main.py]: Assertions in method is also strict. If user does not provide the condition, just skip it will be OK

**Review by onesuper — APPROVED** [review]

File: datawaves/ethereum/condition.py [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+
+def _return_self_condition(alias: str, _type: str) -> Condition[str]:
+    return Condition(key_alias=alias, key_type=_type, transform=lambda x: x)
+
+
+def _return_hash_condition(alias: str, _type: str) -> Condition[str]:
+    return Condition(key_alias=alias, key_type=_type, transform=lambda x: abs(hash_unsafe_bytes(x)) % 10)
+
+
+def new_trace_conditions():
```
Comment (raycursive) [inline datawaves/ethereum/condition.py:15]: maybe this can be exposed in datawaves/__init__.py?

**Pull request merged**